src/wpublish/QueriesAndChunks.cc

```diff
@@ -74,15 +79,34 @@ nlohmann::json QueryStatistics::getJsonHist() const {
     return js;
 }
 
-/// Return a json object containing information about all tasks.
-/// This can return a very large object and should be used sparingly.
-nlohmann::json QueryStatistics::getJsonTasks() const {
+nlohmann::json QueryStatistics::getJsonTasks(wbase::TaskSelector const& taskSelector) const {
     lock_guard<mutex> const guard(_qStatsMtx);
-    nlohmann::json js = nlohmann::json::array();
-    for (auto&& task : _tasks) {
-        js.push_back(task->getJson());
+    nlohmann::json result = nlohmann::json::object();
+    result["snapshotTime_msec"] = util::TimeUtils::now();
+    result["entries"] = nlohmann::json::array();
+    auto& taskEntriesJson = result["entries"];
+    uint32_t numTasksSelected = 0;
+    if (taskSelector.includeTasks) {
+        auto const& ids = taskSelector.queryIds;
+        auto const& states = taskSelector.taskStates;
+        for (auto&& task : _tasks) {
+            bool const selectedByQueryId =
+                    ids.empty() || find(ids.begin(), ids.end(), task->getQueryId()) != ids.end();
+            bool const selectedByTaskState =
+                    states.empty() || find(states.begin(), states.end(), task->state()) != states.end();
+            if (selectedByQueryId && selectedByTaskState) {
+                ++numTasksSelected;
+                // Stop reporting tasks if the limit has been reached. Just keep
+                // counting the number of tasks selected by the filter.
+                if ((taskSelector.maxTasks == 0) || (numTasksSelected < taskSelector.maxTasks)) {
+                    taskEntriesJson.push_back(task->getJson());
```

Comment:
Sending back incomplete lists of tasks could have pretty serious consequences for the displayed dashboard statistics, which could be extremely confusing. There should probably be some indication on the dashboard that it is displaying data for incomplete lists of tasks.

Reply:
Yes, this has already been implemented exactly as you say. I will demo the new version of the Dashboard at the group meeting.